go.mod

```diff
@@ -4,7 +4,17 @@ go 1.14
 
 require (
 	github.com/dgrijalva/jwt-go v3.2.0+incompatible
+	github.com/golang/protobuf v1.4.2 // indirect
 	github.com/grpc-ecosystem/go-grpc-middleware v1.2.2
+	golang.org/x/oauth2 v0.0.0-20200107190931-bf48bf16ab8d // indirect
+	google.golang.org/genproto v0.0.0-20200806141610-86f49bd18e98 // indirect
 	google.golang.org/grpc v1.34.0
-	google.golang.org/grpc/examples v0.0.0-20201209011439-fd32f6a4fefe // indirect
+	google.golang.org/protobuf v1.25.0 // indirect
+)
+
+replace (
+	// Avoid this PR that is importing bleeding edge grpc https://github.com/grpc-ecosystem/go-grpc-middleware/commit/5be27de402455e9e74d35ae771c2f4ee983dc726#diff-e72cb4d89d89634404b11bf1d01ac46aafdf8f38463374922ee81142fa5d9606
+	github.com/grpc-ecosystem/go-grpc-middleware => github.com/grpc-ecosystem/go-grpc-middleware v1.2.0
+	// grpc v1.30+ broke many api contracts, the world is not ready for it
+	google.golang.org/grpc => google.golang.org/grpc v1.29.1
```

Comment:
I wonder if this is really the best tool to solve the problem... This won't allow apps to upgrade to >1.34 when it's available. If a much lower version were used in `require`, would this work better with the minimal version selection algorithm?

Reply:
go build,test will upgrade versions found in go.mod. The problem is that grpc and go-grpc-middleware didn't honor backwards compatibility and broke API (go build) in a minor release.

Reply:
replace isn't honored when module is required by another module. So this just ensures that this git repo builds correctly https://github.com/golang/go/wiki/Modules#gomod

Reply:
> go build,test will upgrade versions found in go.mod.

This isn't quite how it works. `go build,test` will edit go.mod with the _minimum_ version required to support the dependencies. For example, in this go.mod, if you set `require google.golang.org/grpc v1.15.0` (the first version that supports modules), it will not automatically upgrade to the latest grpc. It will find that [go-grpc-middleware v1.2.2 (which is also required) requires v1.29.1 ](https://github.com/grpc-ecosystem/go-grpc-middleware/blob/v1.2.2/go.mod#L22), so it will set `require google.golang.org/grpc v1.29.1`.

https://golang.org/ref/mod#:~:text=The%20main%20module%20requires%20module,1.3%20and%20C%201.4%2C%20respectively.

IMO, a generic module like this should `require` only the minimum versions needed for the functions it calls, to support the widest set of user versions. For this repo, it only calls `metadata.NewIncomingContext()`, which has been unchanged for ages, so there's no benefit to requiring grpc 1.34.0.

> replace isn't honored when module is required by another module.

This sounds like it will cause problems here. If v1.34.0 is required, and `replace` is ignored, users of this module are going to have to use their own `replace` directive instead of being able to simply get v1.29.1 via MVS.

I'd recommend downgrading the `require` directives in this repo as much as possible for the tests to still pass so that usage can tolerate a wide variety of newer versions.

Reply:
I removed the replace, ran go mod tidy and it added grpc v1.34.0 and grpc/examples which will break most applications. I don't think the minimum api verification works on the grpc module.